Use semantic heading structure on course results page

hansschuijff opened the pull request:
Fixes #3200 

### Changes proposed in this Pull Request
* Fixes Non-semantic use of headings on course results page.
* Wraps module-name in `<h3>` tag.
* Wraps lesson-title in `<h4>` tag when in modules, otherwise `<h3>` tag.
* Adjusts scss/css to keep grades floating to right.

### Testing instructions
* Impacts course results pages `/course/{course-slug}/results/`

### Changed Template
- `templates/course-results/lessons.php` - Use semantic heading structure

donnapep commented on assets/css/frontend.scss:
```diff
@@ -690,7 +690,7 @@ section.entry span.course-lesson-progress { margin-left: 10px; }
 /* Course Results */
 .course-results-lessons {
-  h2 {
+  h2, h3, h4 {
```
This SCSS can be cleaned up to reduce bloat:

- If we're no longer using an `h2`, that selector can be removed.
- Since the `total-grade` element is always inside an `h3`, the CSS for it doesn't need to be applied to the `h4` as well.

The resulting SCSS would be something like this:
```
/* Course Results */
.course-results-lessons {
  h3, h4 {
    margin: 20px 0;

    .lesson-grade {
      float: right;
    }
  }

  h3.total-grade {
    text-decoration: underline;

    .lesson-grade {
      text-decoration: underline;
    }
  }
}
```

hansschuijff replied:
The h2 is still applied to the "Lessons" heading, so perhaps the margin should stay for that?
`  h2 {
    margin: 20px 0;
  }`

donnapep commented on templates/course-results/lessons.php:
```diff
@@ -82,7 +82,7 @@
 						}
```
Any time a template is changed, the version at the top needs to be updated. This will likely go in version 3.2.0.

donnapep commented on templates/course-results/lessons.php:
```diff
@@ -115,16 +115,12 @@
-
-			</h3>
+				<h3><?php esc_html_e( 'Other Lessons', 'sensei-lms' ); ?></h3>
+			<?php endif; // $course_has_lessons_in_modules ?>
```
Can this comment be removed? We're trying to move away from doing this.

hansschuijff replied:
Good to know.

donnapep commented on templates/course-results/lessons.php:
```diff
@@ -144,7 +140,7 @@
 				?>
 
-				<h2>
+				<?php echo $course_has_lessons_in_modules ? '<h4>' : '<h3>'; ?>
```
The `?>` on line 141 could be removed so that the PHP can continue without needing to add `<?php` again.

donnapep commented on templates/course-results/lessons.php:
```diff
@@ -115,16 +115,12 @@
 				?>
-
-			</h3>
+				<h3><?php esc_html_e( 'Other Lessons', 'sensei-lms' ); ?></h3>
```
This should be at the same level as the "Lessons" header, so an `h2`.

hansschuijff replied:
Are you sure? I don't think so. It was already coded as a module-heading for lessons that remain after the modules are all listed. Something like a module containing the left-over lessons, just in case. Making this a h2 would impact the lessons heading level too, but would also visually have impact, since it would make the other lessons presented as big as the module-headings, making the presentation a bit confusing. H3 makes more sense here.

Also when we have a heading "Lessons" What would be the meaning of a same-level "Other lessons"? What lessons are excluded from "Lessons"?

donnapep replied:
Yes, I'm sure. They are rendered at the same level on the single course page, so we should be consistent here.

donnapep replied:
No, I'm not sure anymore. 😄  I'll circle back around to this.

hansschuijff replied:
Hi @donnapep, what you ask would be a small change for me to make, but please consider the context first. I have compared the two pages and what you are proposing is either not enough or not the correct choice here. If I know what you want with that, I will be happy to change what I can.

You are right that in the course page a h2 is used, but h2 in the single course page is the same level as the module heading. Here the module level heading is a h3. Considering the other elements in the list, changing the other lessons tag alone is not enough and it should remain h3 or all other elements should change with it. See below for the outline html.

I've compared the two pages and found they are really not that comparable in how they build the lists. On the course page there are two lists separating lessons outside modules from lessons in modules and the lesson list there is actually a list (ul) with list-items instead of a sequence of headers. 

That is actually a better choice, since having only a sequence of headings isn't really conforming to the html5 standard. A heading should be followed by the content of that heading, not being just another heading.

In the course page the first list has the headline "Modules" and the second list has the headline "Other lessons". The modules are each wrapped in an article tag and within the article there is further structuring. Between the module heading and the list of lessons there is still another heading "lessons", that is not part of the course results page, since there the main heading is already "lessons".

The course results page presents just one combined list ( although it isn't really a list, in html sense, just a sequence of headings ). There is one article wrapper for the entire list, so not like the course page lesson list too. The heading "Lessons" contains both all modules and lessons and ends with a heading containing the total grade. Quite different beasts, so to say.

Also the course page does a much better job in styling/presenting the lists and also the html structure seems better. Making the lists really consistent with each other on a html/structure-level would be a larger change than what is proposed here and I don't know if we want to do that here and now.

The course page uses much more elements in building the list. Using the article tag means that the heading in the article in theory may have a separate structure from their container html, in which they can even have a h1 per article element, since an article tag is meant as being an article by itself and the main heading would then be the title of the article. But it doesn't use h1, it uses h2, not really following the outline of headings by itself, but also not starting completely fresh. On the course page that is solved by a better styling, so it presents quit neat, but that wasn't really done on the course results page.

I'm not sure how search engine handle that at this point, since for long browsers and seo outlined ignoring the section and article tags, just based on the h-level. In that case the course page may not be structured well too.

Here the resulting html outline of both pages. See here too how just making other lessons a h2 would impact the outline of the page and not be semantic. Note that there is even a class "module-name" in the "Other lessons" h3 tag (which I didn't add to the code) suggesting the intent of making it be a module-headline.

### The single course page
```
<article class="course">
	
	<!-- course content -->

	<!-- First List of lessons per module modules -->
	<header class="modules-title">
		<h2>Modules</h2>
	</header>
	<!-- per module -->
	<article class="module">
	   <header><h2>Module title</h2></header>
	   <section class="entry">
	      <p class="module-description"></p>
	      <section class="module-lessons">
	         <header>
	            <h3>Lessons</h3>
	         </header>
	         <ul class="lessons-list">
				<-- per module lesson -->
	            <li class=""><a href="lesson-url" title="Lesson title">Lesson title</a></li>
	         </ul>
	      </section>
	   </section>
	</article>

	<!-- Second List of lessons that are not part of modules -->
	<header>
		<h2>Other lessons</h2>
	</header>
	<section class="course-lessons">
	   <article class="post-87432 lesson type-lesson status-publish post entry">
	      <header class="lesson-title">
	         <h2><a href="lesson-url" title="Lesson title">Lesson title</a></h2>
	         <p class="lesson-meta"></p>
	      </header>
	      <section class="entry"></section>
	   </article>
	</section>
</article>
```
### The course results page
```
<section class="course-results-lessons">
   <header>
      <h2>Lessons</h2>
   </header>
   <article class="course">

      <!-- per module -->
      <h3 class="module-name">Module title</h3>

      <!-- per lesson in a module -->
      <h4 class="lesson-title">
         <a href="lesson-url" title="Lesson-title">Lesson-title</a>
         <span class="lesson-grade"></span>
      </h4>

      <!-- after last module a block with lessons without modules -->
      <h3 class="module-name">
         other-lessons
      </h3>

      <!-- per lesson not attached to module -->
      <h4 class="lesson-title">
         <a href="lesson-url" title="Lesson-title">Lesson-title</a>
         <span class="lesson-grade">0%</span>
      </h4>

      <!-- ending with total grades -->
      <h3 class="total-grade">
         Totale resultaat			
         <span class="lesson-grade">0%</span>
      </h3>
   </article>
</section>

```

hansschuijff replied:
> No, I'm not sure anymore. 😄 I'll circle back around to this.

Understood. I will wait till you know. Thanks, so far.

donnapep commented on templates/course-results/lessons.php:
```diff
@@ -159,14 +155,14 @@
 					<span class="lesson-grade"><?php echo esc_html( $lesson_grade ); ?></span>
 
-				</h2>
+				<?php echo $course_has_lessons_in_modules ? '</h4>' : '</h3>'; ?>
```
"Other Lessons" are never inside a module, so this should always be an `h3` given the header is always an `h2`. Right now it looks like this:

![Screen Shot 2020-06-12 at 7 17 39 AM](https://user-images.githubusercontent.com/1190420/84497465-d5233d00-ac7c-11ea-853a-47ebcc12bba3.jpg)

hansschuijff commented:
@donnapep Thanks for the review. I processed your change requests. Exceptions: Left h2 in scss for margins only, since that applied to the Lessons heading too, and didn't change heading level of "Other lessons", see my reaction on that.

donnapep commented:
I've audited the single course page and the course results page. The single course page is not structured correctly either, so is not the best to use for a comparison.

Here's what I know:
- The "Lessons" heading is semantically equivalent to the "Other Lessons" heading, and they should also be styled the same.
- The lessons listed under either of these headings should also be styled the same.

Right now, we're debating the use of HTML headings, but there's a second issue related to the styling of these elements. While the headings are representative of the structure, we can and should manipulate the CSS if required, and it seems to be required here.

Based on this, here are the changes still needed for the course results page:
- "Other Lessons" heading changes to an `h2`.
- Lessons under "Other Lessons" change to `h3`.
- _Styling of lessons under "Other Lessons" are altered _to_ look the same as the lessons that are part of a module._

With the above, we both accurately describe the structure of the page, and style the elements accordingly so that the corresponding levels look the same.

hansschuijff commented:
@donnapep I've changed the heading level of other lessons and the lessons under that, to reflect your CR. I also changed the level of the total grade, since it now the lesson list is basically split in two lists and total grade should not be part of only the second list, what would be the case if it remained a h3.

If you want, we could probably better follow the course page structure and headings some more. That means that the first h2 should say Modules like it does in the course page, that is also clearer when there are modules. Making this heading optional and changing "Other Lessons" in "Lessons" when there are no modules, could implement that. In the course page there is a "Lessons" heading within each module and we could follow that too.

Also it is strange that in the h2 Lessons chapter has an article tag with .course within it. It should be the other way around, since lessons exist within a course, not the other way around. Article should wrap the Lessons headline too.

And it would be better to change the tags of lessons to li-items, making headings in the structure more meaningful and it is strange that course results is a heading instead of just a p-item.

donnapep commented:
@hansschuijff I don't think we want to take on a complete revamp of the course results page at this time. This seems like a good intermediate step for now.
